core/lib/vm_interface/src/vm.rs

```diff
@@ -38,67 +32,48 @@ pub trait VmInterface {
         execution_mode: VmExecutionMode,
     ) -> VmExecutionResultAndLogs;
 
-    /// Get bootloader memory.
-    fn get_bootloader_memory(&self) -> BootloaderMemory;
-
-    /// Get last transaction's compressed bytecodes.
-    fn get_last_tx_compressed_bytecodes(&self) -> Vec<CompressedBytecodeInfo>;
-
     /// Start a new L2 block.
     fn start_new_l2_block(&mut self, l2_block_env: L2BlockEnv);
 
-    /// Get the current state of the virtual machine.
-    fn get_current_execution_state(&self) -> CurrentExecutionState;
-
-    /// Execute transaction with optional bytecode compression.
-    fn execute_transaction_with_bytecode_compression(
-        &mut self,
-        tx: Transaction,
-        with_compression: bool,
-    ) -> (
-        Result<(), BytecodeCompressionError>,
-        VmExecutionResultAndLogs,
-    ) {
-        self.inspect_transaction_with_bytecode_compression(
-            Self::TracerDispatcher::default(),
-            tx,
-            with_compression,
-        )
-    }
-
     /// Execute transaction with optional bytecode compression using custom tracers.
     fn inspect_transaction_with_bytecode_compression(
         &mut self,
         tracer: Self::TracerDispatcher,
         tx: Transaction,
         with_compression: bool,
-    ) -> (
-        Result<(), BytecodeCompressionError>,
-        VmExecutionResultAndLogs,
-    );
+    ) -> (BytecodeCompressionResult, VmExecutionResultAndLogs);
 
     /// Record VM memory metrics.
     fn record_vm_memory_metrics(&self) -> VmMemoryMetrics;
 
-    /// How much gas is left in the current stack frame.
-    fn gas_remaining(&self) -> u32;
-
     /// Execute batch till the end and return the result, with final execution state
     /// and bootloader memory.
-    fn finish_batch(&mut self) -> FinishedL1Batch {
-        let result = self.execute(VmExecutionMode::Batch);
-        let execution_state = self.get_current_execution_state();
-        let bootloader_memory = self.get_bootloader_memory();
-        FinishedL1Batch {
-            block_tip_execution_result: result,
-            final_execution_state: execution_state,
-            final_bootloader_memory: Some(bootloader_memory),
-            pubdata_input: None,
-            state_diffs: None,
-        }
+    fn finish_batch(&mut self) -> FinishedL1Batch;
+}
+
+/// Extension trait for [`VmInterface`] that provides some additional methods.
+pub trait VmInterfaceExt: VmInterface {
+    /// Executes the next VM step (either next transaction or bootloader or the whole batch).
```

Comment:
What is the purpose of putting the methods with default implemenations into an extension trait?

Reply:
To improve usage experience without polluting the main trait (and allowing to unexpectedly alter its behavior); it's a common practice, see [FutureExt](https://docs.rs/futures/latest/futures/future/trait.FutureExt.html), [StreamExt](https://docs.rs/futures/latest/futures/stream/trait.StreamExt.html) (note that [Stream](https://docs.rs/futures/latest/futures/stream/trait.Stream.html) is defined within the same crate), [Itertools](https://docs.rs/itertools/latest/itertools/trait.Itertools.html).

Reply:
So that there's no ambiguousness of whether an implementation may implement these methods (e.g., for performance reasons). This, in turn, simplifies reasoning around implementations; note e.g. how I had to include optional methods in the shadow VM in case they were overridden by either of main and shadow VMs. In the general case, an extension trait could be non-object safe even if the extended trait is, but this is not relevant in this case.